### PR TITLE
Personalized recommendations v0

### PR DESCRIPTION
**Pull request by jasnaurbancic**
**Enhancements:**

- added support for personalized recommendations based on previously accessed material
- added support for on-action user model updates
- added import scripts to generate user and material models

### CODE REVIEW

File: src/server/recsys/engine/models/nearest-neighbors.js [hunk 1]
```diff
@@ -94,9 +94,18 @@ class NearestNeighbors {
...
         try {
             // transform the query json into a sparse vector
-            let queryRec = query.url ? 
-                store.recordByName(query.url) :
-                store.newRecord({ description: query.text });
+            // this step here might be tricky!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+            let queryRec;
+            if (query.hasOwnProperty('uuid') && query.hasOwnProperty('wikipediaConceptNames') && 
+                query.hasOwnProperty('wikipediaConceptSupport')) {
+                    queryRec = store.newRecord({uri: query.uuid, title: null, description: null, provider: null,
```
Comment (ErikNovak) [inline src/server/recsys/engine/models/nearest-neighbors.js]: can you put each attribute in its own line?

File: src/server/recsys/engine/models/nearest-neighbors.js [hunk 2]
```diff
@@ -94,9 +94,18 @@ class NearestNeighbors {
 
         try {
             // transform the query json into a sparse vector
-            let queryRec = query.url ? 
-                store.recordByName(query.url) :
-                store.newRecord({ description: query.text });
+            // this step here might be tricky!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
```
Comment (ErikNovak) [inline src/server/recsys/engine/models/nearest-neighbors.js]: is this comment necessary? :)

File: src/server/platform/routes/v1/activity-logging.js [hunk 3]
```diff
@@ -299,6 +300,7 @@ module.exports = function (pg, logger) {
         pg.insert(activity, 'client_activity', (error) => {
             if (error) {
                 // log postgres error
+                console.log(activity);
```
Comment (ErikNovak) [inline src/server/platform/routes/v1/activity-logging.js]: remove console.log

File: src/server/recsys/engine/x5recommend.js [hunk 4]
```diff
@@ -353,5 +412,107 @@ class x5recommend {
...
+                    resolve({error: 'Error fetching user model'});
+                    return;
+                }
+                
+                if (!res || res.length == 0){
+                    resolve({error: 'Cookie is not in the database - unable to fetch the user'});
+                    return;
+                }
+                
+                console.log('Result of pg call: ', res);
```
Comment (ErikNovak) [inline src/server/recsys/engine/x5recommend.js]: remove console.log

File: src/server/recsys/engine/x5recommend.js [hunk 5]
```diff
@@ -353,5 +412,107 @@ class x5recommend {
...
+                
+                res = res[0];
+                
+                console.log(res);
```
Comment (ErikNovak) [inline src/server/recsys/engine/x5recommend.js]: remove console.log

File: src/server/recsys/scripts/load-material-models.js [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+                        if (!maxKey){
+                            maxKey = key;
+                        }
+                        else if (language[key] > language[maxKey]){
+                            maxKey = key;
+                        }
+                    }
+                    language = maxKey;                    
+                    
+                    //console.log(provideruri, type, language, JSON.stringify(wiki));
```
Comment (ErikNovak) [inline src/server/recsys/scripts/load-material-models.js]: remove commented console.log

File: src/server/recsys/test/recommender-tests.js [hunk 7]
```diff
@@ -162,20 +162,29 @@ describe('recommender-tests.js: Recommender engine model unit tests.', function
             done();
         });
 
-        it('Should handle a text and url query as url query', function (done) {
+        /*it('Should handle a text and url query as url query', function (done) {
```
Comment (ErikNovak) [inline src/server/recsys/test/recommender-tests.js]: why is this test commented out?

**Comment by jasnaurbancic** [comment]
Za pognat: Najprej poženeš recsys.js, da se ti naredijo postgres baze, potem ugasneš in poženeš buildRecSys.js, da se ti naredijo QMiner baze. (Delam na združitvi tega dvojega).

**Comment by jasnaurbancic** [comment]
Reverted server starting back to initial method.
Added an import script to create user and material models. Location src/load/recsys-data-import.js.

**Comment by ErikNovak** [comment]
While running `recsys-data-import.js` I get the following error:
```bash
if (fn === null) throw new Error("Callback was already called.");
                         ^

Error: Callback was already called.
    at /home/erikn/Documents/code/x5gon/platform-x5gon/node_modules/async/dist/async.js:955:32
    at /home/erikn/Documents/code/x5gon/platform-x5gon/src/lib/update-user-models.js:109:25
    at Query.client.query (/home/erikn/Documents/code/x5gon/platform-x5gon/src/lib/postgresQL.js:179:42)
    at Query.handleReadyForQuery (/home/erikn/Documents/code/x5gon/platform-x5gon/node_modules/pg/lib/query.js:126:10)
    at Connection.<anonymous> (/home/erikn/Documents/code/x5gon/platform-x5gon/node_modules/pg/lib/client.js:163:19)
    at Connection.emit (events.js:160:13)
    at Socket.<anonymous> (/home/erikn/Documents/code/x5gon/platform-x5gon/node_modules/pg/lib/connection.js:117:12)
    at Socket.emit (events.js:160:13)
    at addChunk (_stream_readable.js:269:12)
    at readableAddChunk (_stream_readable.js:256:11)
```

**Pull request merged**